jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java

```diff
@@ -98,23 +98,49 @@ public int available() throws IOException
     }
 
     @Override
-    public void close()
+    public void close() throws IOException
     {
         // If we have already reached a real EOF or a persistent failure, close is a noop.
         if (chunk == Content.Chunk.EOF || Content.Chunk.isFailure(chunk, true))
             return;
 
+        boolean contentSkipped = false;
+
         // If we have a chunk here, then it needs to be released
         if (chunk != null)
         {
+            contentSkipped = chunk.hasRemaining();
             chunk.release();
+            chunk = Content.Chunk.next(chunk);
+        }
+
+        // If we don't have a chunk and have not skipped content, try one read looking for EOF
+        if (!contentSkipped && chunk == null)
+        {
+            chunk = content.read();
+
+            // If we read a chunk
+            if (chunk != null)
+            {
+                // Handle a failure as read would
+                if (Content.Chunk.isFailure(chunk))
+                {
+                    Content.Chunk c = chunk;
+                    chunk = Content.Chunk.next(c);
+                    throw IO.rethrow(c.getFailure());
+                }
```

Comment:
I would not do this, as this is a close() and I don't want an exception in my face if I can't read -- it's a quasi-normal behavior if the other side has already closed as well.
I'd just store the failure (in case of weird extra reads) and return.